app/components/projects/settings/life_cycles/step_component.rb

```diff
@@ -0,0 +1,53 @@
+# -- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2010-2024 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+
+module Projects
+  module Settings
+    module LifeCycles
+      class StepComponent < ApplicationComponent
+        include ApplicationHelper
+        include OpPrimer::ComponentHelpers
+        include OpTurbo::Streamable
+
+        options :definition,
+                :project
+
+        def active_in_project?
+          project
+            .life_cycle_steps
+            .detect { |project_lc| project_lc.definition_id == definition.id }
+            &.active
+        end
```

Comment:
Looks like I forgot to write about this - feels not good that all steps will be iterated for every step/row, maybe better to find definitions and matching steps in index component, and pass both to the every step component instead? But not sure if it will be better

Reply:
The problem here is that it is not guaranteed that a corresponding `life_cycle_step` exists for every definition. We don't seed them on project & definition creation. Otherwise it would be a good approach to work through that list instead of through the definitions.   

An approach would be to build the non existing steps in the index action and then pass that step into the step component. Would definitely improve the interface as only the step would have to be passed into the `StepComponent`. The `definition` could then be accessed from the step within the component. This all comes at the cost of increased complexity in the index component.

Performancewise, I'd not expect it to be better though. It exchanges a non optimal loop for AR model instantiation. As there are probably not many steps, rather sure there are less than 20 and very sure there are less than 100, I would also expect the impact of the non optimal loop to be minimal.

Reply:
Absent steps can be also simply nils, or only boolean for active state can be passed.

```rb
def life_cycle_definitions_and_step_active
  active_ids = project.life_cycle_steps.where(active: true).pluck(:definition_id).to_set

  life_cycle_definitions.all.map do |definition|
    [definition.id, definition.id.in?(active_ids)]
  end
end

# …

life_cycle_definitions_and_step_active.each do |definition, active|
  # …
  Projects::Settings::LifeCycles::StepComponent.new(project:, definition:, active?: active))
  # …
end
```

Reply:
Moved it. I didn't need the project in the step component so the change is one parameter for one which is nice.